### DISCUSSION TITLE
C++ code generator for Xstate released

### DISCUSSION THREAD

**Discussion by shuvalov-mdb** [statelyai/xstate#1608] [Show and tell]
I decided to provide Xstate goodies to server-side C++ developers with C++ generator, ready for production.
Essentially it's a TypeScript library with templates, sourcing Xstate Machine as data model to generate C++.
Github: [https://github.com/shuvalov-mdb/xstate-cpp-generator](https://github.com/shuvalov-mdb/xstate-cpp-generator)
NPM package: [https://www.npmjs.com/package/xstate-cpp-generator](https://www.npmjs.com/package/xstate-cpp-generator)

Please post here what you think, I don't have any other message board.

**Comment by davidkpiano** [comment]
This is really awesome, I'm going to have to play around with it! I added it [to the resources page](https://xstate.js.org/docs/about/resources.html#articles-and-videos).

**Reply by shuvalov-mdb** [reply]
Thank you!

**Reply by agorikov** [reply]
Awesome work!

I have few questions, regarding using code generator path for devops task:
* Is it (easily) achievable to plug C++ generated code into golang? Assuming that side effects (entry, exit, on transition) to be implemented in golang.
* Timers in C++ template, has comments that they start new thread and put it on pause. However, XState itself runs single-threaded. What was motivation behind introducing multi-threaded (and locks) into xstate-cpp-generator?
* How much effort would be to do similar generator for golang?
* Could generated code connect to xstate inspector, same as XState do from browser?
